src/util/futurepool.rs

```diff
@@ -0,0 +1,259 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// This mod implemented a wrapped future pool that supports `on_tick()` which is driven by
+/// tasks and is invoked no less than the specific interval.
+
+use std::fmt;
+use std::cell::{Cell, RefCell, RefMut};
+use std::sync::{mpsc, Arc};
+use std::thread;
+use std::time::Duration;
+use futures::Future;
+use futures_cpupool as cpupool;
+
+use util;
+use util::time::Instant;
+use util::collections::HashMap;
+
+pub trait Context: fmt::Debug + Send + 'static {
+    fn on_tick(&mut self) {}
+}
+
+#[derive(Debug)]
+/// A delegator to wrap `Context` to provide `on_tick` feature.
+struct ContextDelegator<T: Context> {
+    tick_interval: Duration,
+    /// TODO: Can be replace by `UnsafeCell` to eliminate the runtime cost of borrow checking.
+    inner: RefCell<T>,
+    last_tick: Cell<Option<Instant>>,
+}
+
+impl<T: Context> ContextDelegator<T> {
+    fn new(context: T, tick_interval: Duration) -> ContextDelegator<T> {
+        ContextDelegator {
+            tick_interval,
+            inner: RefCell::new(context),
+            last_tick: Cell::new(None),
+        }
+    }
+
+    fn get_context(&self) -> RefMut<T> {
+        self.inner.borrow_mut()
+    }
+
+    fn on_task_finish(&self) {
+        let now = Instant::now_coarse();
+        let last_tick = self.last_tick.get();
+        if last_tick.is_none() {
```

Comment:
Use `if let` pattern.

Reply:
`last_tick` is used [here](https://github.com/pingcap/tikv/pull/2699/files/4c0b41348ba46f8a381b78e27c3682a4d494d747#diff-1e32e8866f7a4a827894089d8a975714R63) as well. I'm afraid that we cannot use `if let` syntax here. In fact the `last_tick` variable is used to reduce one call to `.get()`.

Reply:
you can write code like
```rust
        if let Some(last_tick) = self.last_tick.get() {
            if now.duration_since(last_tick) < self.tick_interval {
                return;
            }
            self.get_context().on_tick();
        }
        self.last_tick.set(Some(now));
```

Reply:
Wow. I prefer current one, which have less if- levels, easier to understand its logic.